src/modules/roc_core/target_darwin/roc_core/time.cpp

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2015 Mikhail Baranov
+ * Copyright (c) 2015 Victor Gaydov
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <errno.h>
+#include <sys/time.h>
+#include <time.h>
+#include <unistd.h>
+
+#include "roc_core/panic.h"
+#include "roc_core/time.h"
+
+namespace roc {
+namespace core {
+
+#include <mach/clock.h>
+#include <mach/clock_types.h>
+#include <mach/mach.h>
+#include <mach/mach_time.h>
+
+/* As Apple mentioned: "The mach_timespec_t API is deprecated in OS X. The
+ * newer and preferred API is based on timer objects that in turn use
+ * AbsoluteTime as the basic data type".
+ *
+ * We still use a function from the old OS X API (clock_sleep), because OS X
+ * API only provides one method to wait until some period (mach_wait_until)
+ * and doesn't allow to select a clock against which the sleep interval is to
+ * be measured to specify the sleep interval as either an absolute or a
+ * relative value.
+ *
+ * https://developer.apple.com/library/content/documentation/Darwin/Conceptual/KernelProgramming/Mach/Mach.html#//apple_ref/doc/uid/TP30000905-CH209-TPXREF111
+ */
+uint64_t timestamp_ms() {
+    /* mach_absolute_time() returns a Mach Time unit - clock ticks. The
+     * length of a tick is a CPU dependent. On most Intel CPUs it probably
+     * will be 1 nanoseconds per tick, but let's not rely on this. Mach
+     * provides a conversation factor that can be used to convert abstract
+     * mach time units to nanoseconds.
+     */
+    static double steady_factor = 0;
+    static uint64_t tm_start = 0;
+
+    if (!tm_start) {
```

Comment:
This is not threadsafe. You could either use a mutex with a double-check (so only the first calls will be expensive), or a singleton.